File: cartesia-vst-ui/src/engine/engine.ts
/** Matilda v2 stepping engine — port of cartesia/engine.py */

import { advancePath, pathXY } from "./movement";
import type { Cell, Layer, Patch } from "./model";
import { mapDegree, scaleOffsets } from "./pitch";

export type StepEvent = {
  degree: number | null;
  velocity: number | null;
  x: number;
  y: number;
  layerIdx: number;
  stepIndex: number;
  fired: boolean;
};

/** Merge UI patch edits without clobbering live path / scheduler state. */
export function syncPatchFromUI(engine: CartesiaEngine, next: Patch): void {
  const ep = engine.patch;
  ep.title = next.title;
  ep.root = next.root;
  ep.mode = next.mode;
  ep.quantize = next.quantize;
  ep.minOctave = next.minOctave;
  ep.maxOctave = next.maxOctave;
  ep.masterDivision = next.masterDivision;
  ep.playMode = next.playMode;
  ep.playOnTransport = next.playOnTransport;
  ep.selectedLayer = next.selectedLayer;
  ep.polyVoices = next.polyVoices;
  ep.seed = next.seed;

  for (let i = 0; i < 4; i++) {
    ep.layers[i].active = next.layers[i].active;
    ep.layers[i].movement = next.layers[i].movement;
    ep.layers[i].randomSkipProb = next.layers[i].randomSkipProb;
    for (let y = 0; y < 4; y++)
      for (let x = 0; x < 4; x++)
        Object.assign(ep.layers[i].cells[y][x], next.layers[i].cells[y][x]);
  }

  engine.reconcilePlayingLayer();
}

export class CartesiaEngine {
  static readonly STEPS_PER_LAYER_PASS = 16;

  patch: Patch;
  masterTick = 0;
  lastStepIndex = -1;
  lastPlayingLayer = 0;
  lastFired = false;

  /** Pinned layer index (0…3) — only changes after a 16-step pass or forced skip. */
  private playingLayerIdx = 0;
  private stepsOnLayer = 0;
  private rngState: number;

  constructor(patch: Patch) {
    this.patch = patch;
    this.rngState = patch.seed ?? Date.now() & 0xffff;
    this.playingLayerIdx = CartesiaEngine.firstActiveLayer(patch);
  }

  static firstActiveLayer(patch: Patch): number {
    const idx = patch.layers.findIndex(l => l.active);
    return idx >= 0 ? idx : 0;
  }

  reset(): void {
    this.masterTick = 0;
    this.stepsOnLayer = 0;
    this.playingLayerIdx = CartesiaEngine.firstActiveLayer(this.patch);
    this.lastStepIndex = -1;
    this.lastFired = false;
    for (const layer of this.patch.layers) {
      layer.path.stepIndex = 0;
      layer.path.direction = 1;
      layer.path.atEndHold = 0;
      layer.path.randomBag = [];
      layer.path.randomBagPos = 0;
      layer.stepIndex = 0;
      layer.stepDir = 1;
    }
  }

  /** If the pinned layer was deactivated mid-play, skip to the next active layer. */
  reconcilePlayingLayer(): void {
    if (this.patch.layers[this.playingLayerIdx]?.active) return;
    this.playingLayerIdx = this.findNextActiveLayer(this.playingLayerIdx);
    this.stepsOnLayer = 0;
  }

  /** Next active layer in fixed order 0 → 1 → 2 → 3, wrapping. */
  findNextActiveLayer(fromLayer: number): number {
    for (let i = 1; i <= 4; i++) {
      const idx = (fromLayer + i) % 4;
      if (this.patch.layers[idx].active) return idx;
    }
    return 0;
  }

  private playingLayerIndex(): number {
    return this.playingLayerIdx;
  }

  private rng(): number {
    let t = (this.rngState += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  private chance(prob: number): boolean {
    return this.rng() < Math.max(0, Math.min(1, prob));
  }

  private effectiveTriggerProb(cell: Cell): number {
    return cell.triggerArmed ? Math.max(0, Math.min(1, cell.triggerProb)) : 1;
  }

  private applyJitter(semitone: number, cell: Cell): number {
    if (!cell.jitterArmed || cell.jitterAmount <= 0) return semitone;
    const span = this.patch.quantize
      ? Math.max(1, scaleOffsets(this.patch.mode).length)
      : 12;
    const maxDelta = Math.max(1, Math.round(cell.jitterAmount * span));
    const delta = Math.floor(this.rng() * (2 * maxDelta + 1)) - maxDelta;
    return semitone + delta;
  }

  private resolveCell(layer: Layer, x: number, y: number): { fired: boolean; degree: number | null; velocity: number | null } {
    const cell = layer.cells[y][x];
    if (!cell.gate) return { fired: false, degree: null, velocity: null };
    if (!this.chance(this.effectiveTriggerProb(cell))) {
      return { fired: false, degree: null, velocity: null };
    }
    let semitone = mapDegree(cell.degree, this.patch);
    semitone += cell.octaveOffset * 12;
    semitone = this.applyJitter(semitone, cell);
    return { fired: true, degree: semitone, velocity: cell.velocity };
  }

  private advanceLayerPath(layer: Layer): void {
    advancePath(layer.movement, layer.path, {
      randomSkipProb: layer.randomSkipProb,
      rng: () => this.rng(),
    });
    layer.stepIndex = layer.path.stepIndex;
    layer.stepDir = layer.path.direction;
  }

  private maybeSwitchLayer(): void {
    this.stepsOnLayer += 1;
    if (this.stepsOnLayer < CartesiaEngine.STEPS_PER_LAYER_PASS) return;
    this.stepsOnLayer = 0;
    this.playingLayerIdx = this.findNextActiveLayer(this.playingLayerIdx);
  }

  tick(): StepEvent {
    this.masterTick += 1;
    this.reconcilePlayingLayer();
    const layerIdx = this.playingLayerIndex();
    const layer = this.patch.layers[layerIdx];
    const stepIndex = layer.path.stepIndex;
    const [x, y] = pathXY(layer.path);

    const { fired, degree, velocity } = this.resolveCell(layer, x, y);

    this.lastStepIndex = stepIndex;
    this.lastPlayingLayer = layerIdx;
    this.lastFired = fired;

    this.advanceLayerPath(layer);
    this.maybeSwitchLayer();

    return { degree, velocity, x, y, layerIdx, stepIndex, fired };
  }

  playhead(): { layerIdx: number; stepIndex: number } {
    const layerIdx = this.playingLayerIndex();
    return { layerIdx, stepIndex: this.patch.layers[layerIdx].path.stepIndex };
  }
}

File: cartesia/engine.py
"""Matilda v2 stepping engine — sequential layers + path movement."""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import List, Optional, Tuple

from isobar import Key

from cartesia.model import Layer, Patch
from cartesia.movement import MovementMode, advance_path


SCALE_OFFSETS = {
    "major": [0, 2, 4, 5, 7, 9, 11],
    "minor": [0, 2, 3, 5, 7, 8, 10],
    "dorian": [0, 2, 3, 5, 7, 9, 10],
    "phrygian": [0, 1, 3, 5, 7, 8, 10],
    "lydian": [0, 2, 4, 6, 7, 9, 11],
    "mixolydian": [0, 2, 4, 5, 7, 9, 10],
    "locrian": [0, 1, 3, 5, 6, 8, 10],
    "harmonic_minor": [0, 2, 3, 5, 7, 8, 11],
    "melodic_minor": [0, 2, 3, 5, 7, 9, 11],
    "pentatonic": [0, 2, 4, 7, 9],
    "pentatonic_minor": [0, 3, 5, 7, 10],
    "blues": [0, 3, 5, 6, 7, 10],
    "ionian": [0, 2, 4, 5, 7, 9, 11],
    "aeolian": [0, 2, 3, 5, 7, 8, 10],
}


@dataclass
class StepEvent:
    degree: Optional[int]
    velocity: Optional[int]
    x: int
    y: int
    z: int
    fired: bool
    note: Optional[str] = None


class CartesiaEngine:
    """v2 engine: one path step per master tick; layers play sequentially."""

    STEPS_PER_LAYER_PASS = 16

    def __init__(self, patch: Patch):
        self.patch = patch
        self.key = Key(patch.root, patch.mode)
        mode_key = patch.mode.lower()
        self.scale = SCALE_OFFSETS.get(mode_key, SCALE_OFFSETS["major"])
        self.master_tick = 0
        self._playing_layer_idx = self._first_active_layer()
        self._steps_on_layer = 0
        self._cached_primary: Optional[StepEvent] = None
        self._fields_read = 0
        self._rng = random.Random(patch.seed)

    def _first_active_layer(self) -> int:
        for i, layer in enumerate(self.patch.layers):
            if layer.active:
                return i
        return 0

    def _find_next_active_layer(self, from_layer: int) -> int:
        for i in range(1, 5):
            idx = (from_layer + i) % 4
            if self.patch.layers[idx].active:
                return idx
        return 0

    def _reconcile_playing_layer(self) -> None:
        if self.patch.layers[self._playing_layer_idx].active:
            return
        self._playing_layer_idx = self._find_next_active_layer(self._playing_layer_idx)
        self._steps_on_layer = 0

    def current(self) -> StepEvent:
        if self._cached_primary is None:
            self._do_step()
        return self._cached_primary

    def mark_field_read(self) -> None:
        self._fields_read += 1
        if self._fields_read >= 3:
            self._fields_read = 0
            self._cached_primary = None

    def reset(self) -> None:
        self.master_tick = 0
        self._playing_layer_idx = self._first_active_layer()
        self._steps_on_layer = 0
        for layer in self.patch.layers:
            layer.path.step_index = 0
            layer.path.direction = 1
            layer.path.at_end_hold = 0
            layer.path.random_bag.clear()
            layer.path.random_bag_pos = 0
            layer.step_index = 0
            layer.step_dir = 1
        self._cached_primary = None

    def _playing_layer_index(self) -> int:
        return self._playing_layer_idx

    def _chance(self, prob: float) -> bool:
        return self._rng.random() < max(0.0, min(1.0, prob))

    def _map_degree(self, raw: int) -> int:
        if self.patch.quantize:
            span = max(1, len(self.scale))
            return self.scale[raw % span]
        return raw

    def _apply_jitter(self, semitone: int, cell) -> int:
        if not cell.jitter_armed or cell.jitter_amount <= 0:
            return semitone
        span = max(1, len(self.scale)) if self.patch.quantize else 12
        max_delta = max(1, int(round(cell.jitter_amount * span)))
        delta = self._rng.randint(-max_delta, max_delta)
        return semitone + delta

    def _resolve_cell(self, layer: Layer, x: int, y: int) -> Tuple[Optional[int], Optional[int], bool]:
        cell = layer.get(x, y)
        if not cell.gate:
            return None, None, False
        if not self._chance(cell.effective_trigger_prob()):
            return None, None, False

        semitone = self._map_degree(cell.degree)
        semitone += cell.octave_offset * 12
        semitone = self._apply_jitter(semitone, cell)
        return semitone, cell.velocity, True

    def _advance_layer_path(self, layer: Layer) -> None:
        advance_path(
            layer.movement,
            layer.path,
            random_skip_prob=layer.random_skip_prob,
            rng=self._rng,
        )
        layer.step_index = layer.path.step_index
        layer.step_dir = layer.path.direction

    def _maybe_switch_layer(self) -> None:
        self._steps_on_layer += 1
        if self._steps_on_layer < self.STEPS_PER_LAYER_PASS:
            return
        self._steps_on_layer = 0
        self._playing_layer_idx = self._find_next_active_layer(self._playing_layer_idx)

    def step(self) -> List[StepEvent]:
        self._do_step()
        return [self._cached_primary]

    def _do_step(self) -> List[StepEvent]:
        self.master_tick += 1
        self._reconcile_playing_layer()
        layer_idx = self._playing_layer_index()
        layer = self.patch.layers[layer_idx]
        x, y = layer.path.xy()

        degree, velocity, fired = self._resolve_cell(layer, x, y)

        self._advance_layer_path(layer)
        self._maybe_switch_layer()

        event = StepEvent(
            degree=degree,
            velocity=velocity,
            x=x,
            y=y,
            z=layer_idx,
            fired=fired,
        )
        self._cached_primary = event
        return [event]

    def preview_steps(self, n: int = 16) -> List[StepEvent]:
        clone = CartesiaEngine(self.patch)
        clone.reset()
        out: List[StepEvent] = []
        for _ in range(n):
            out.append(clone._do_step()[0])
        return out

    def playhead(self) -> Tuple[int, int, int]:
        layer_idx = self._playing_layer_index()
        layer = self.patch.layers[layer_idx]
        x, y = layer.path.xy()
        return x, y, layer_idx
